### ISSUE TITLE
Unable to find valid certification path 

### ISSUE THREAD

**Issue by nboissel** [AdoptOpenJDK/openjdk-support#13]
This issue appears on Docker official image (tried with AdoptOpenJDK 8 and 11 in hostpot/amd64 flavors) but I think it can be reproduced on any AdoptOpenJDK Hotspot.

I get the usual PKIX error each time I request some APIs
`Error : javax.ws.rs.ProcessingException: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`
It wouldn't be a problem if this was an auto-signed certificate but it appears on all french government open APIs I tested (for example https://entreprise.data.gouv.fr/api/sirene/v1/siret/<SIRET>).
Those certificates are signed by the same root authority CN = Certigna, O = Dhimyotis, C = FR which seems legit. 
It works on OpenJDK so I imagine it is "just" a difference in root certificate list. 

Was this root certificate voluntarily invalidated? Is it just an omission ?
I know I can add this certificate manually using keystore but I was wondering if this should not be integrated by default.

**Comment by ggam** [comment]
I'm facing the same issue on JDK 13:

```
openjdk version "13" 2019-09-17
OpenJDK Runtime Environment AdoptOpenJDK (build 13+33)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 13+33, mixed mode, sharing)
```

**Comment by ggam** [comment]
Just found in *my case* it was my fault. OpenLiberty doesn't use JDK cacerts by default. I needed to add the following to the server.xml file:
```
<ssl id="defaultSSLConfig" trustStoreRef="myTrustStore"/>

<keyStore id="myTrustStore" location="${java.home}/lib/security/cacerts" password="changeit" />
```

**Comment by tresf** [comment]
Is there a chance that it's due to the SHA-1 signature in the certificate chain per  https://openjdk.java.net/jeps/288?

Quoting:
> Improve the security configuration of the JDK by providing a more flexible mechanism to **disable X.509 certificate chains with SHA-1 based signatures.**

Per https://www.sslshopper.com/ssl-checker.html?hostname=+https%3A%2F%2Fentreprise.data.gouv.fr:

![image](https://user-images.githubusercontent.com/6345473/77506980-9def7d80-6e3d-11ea-9113-c0f43c7f8863.png)

**Comment by aahlenst** [comment]
Closing due to inactivity.

**Comment by tresf** [comment]
Some conversation over here reads very similar: https://github.com/AdoptOpenJDK/openjdk-build/issues/676, but looking at the `cacerts` file, where should this be?

For example, if I cd to `JAVA_HOME/lib/security` and type this command, I get quite a few certs... 

```bash
echo "changeit" | keytool -list -v -keystore cacerts |grep -i VeriSign
```

... however when I try the same command with Certigna, I get nothing

```bash
echo "changeit" | keytool -list -v -keystore cacerts |grep -i Certigna
```

... yet Chrome trusts Certigna as a valid, Root Certificate Authority.

<img src="https://user-images.githubusercontent.com/6345473/80735639-fcfc7e00-8ade-11ea-8012-4fd8ce0dabb2.png" width="400px"/>

**Comment by tresf** [comment]
Related: https://github.com/habitat-sh/core-plans/issues/1488#issue-318362015.

Quoting:

> I'm currently facing a case where I need to access an url validated by a root certificate **Certigna not included in the JRE but trusted by Chrome and Firefox.**

**Comment by aahlenst** [comment]
The cacert file bundled with OpenJDK does not match the list of accepted CAs that is maintained by Mozilla. Historically, when only Oracle was in charge, they were significantly more forgiving than the browser vendors if I remember correctly.

Personally, I hate the fact that our cacert does not match Mozilla's. I think there was a discussion on our Slack but Martijn wanted to discuss that privately because he considered it to be a sensitive security issue (I don't). Hence I don't know what our thinking is here.

As a side note, there's https://github.com/AdoptOpenJDK/openjdk-installer/issues/105. The gist is: At least when you install the Linux packages, Java should either use the system's certificate store or get a cacert generated from Mozilla's. I didn't get around implementing the necessary tooling in Java, yet. If one of you wants to help out, please get in touch.

**Comment by tresf** [comment]
I've worked with Firefox's `certdb` format a bit, but I've found the command-line tools required for interacting with it aren't available by default with certain Linux distributions.

Whereas you're not happy with the fact that `cacert` differs from Mozilla's, I've grown equally frustrated with Mozilla shipping it's own as it causes similar issues to this, but from a browser perspective (fortunately Mozilla recently allowed enterprise policies to use the system's store, but that's a completely different subject).

Anyway, I vaguely remember a property for Windows called `-Djavax.net.ssl.trustStoreType=WINDOWS-ROOT`, is there such a flag for Linux?  Linux seems to handle CA Certs a bit differently than Windows and Mac, so I'm not sure what -- if any -- would be a sane default.  I guess importing them at install time would suffice, but that script would need to re-run on each update (or better, any time the Firefox store is updated), no?

**Comment by tresf** [comment]
> he considered it to be a sensitive security issue (I don't)

When [20,000 certificates are revoked](https://gizmodo.com/23-000-https-certificates-pulled-after-ceo-sends-privat-1823444625), it's public knowledge these days, and issuance is as quick as running `certbot` on a host, I tend to agree.  These things should be in the open, especially when CAs trusted by major browsers are already trusting them.

Of course, the most secure system is one which trusts no CAs at all, but that would be a difficult selling point. :D

**Comment by aahlenst** [comment]
To verify whether the cacert file is really the culprit here:

1. Build a bundle with https://raw.githubusercontent.com/curl/curl/master/lib/mk-ca-bundle.pl.
2. Create the cacert with https://github.com/alastairmccormack/keyutil.
3. Replace the bundled one with the generated file.
4. Test.

If someone could do that, I‘d be very grateful.

@tresf Debian has tooling to create a cacert file from the system‘s keystore. There is a hook system that updates the cacert every time the system‘s keystore is changed. lib/security/cacert is actually a symlink to that file. My plan is to create Java-based tools that replicate that on Red Hat and friends as well as all Debian derivatives. So you have an up-to-date CA list according to your needs that matches the behavour of your system. In case of Debian, it‘s identical with Mozilla‘s list.

**Comment by tresf** [comment]
@aahlenst Using the perl script, I was able to create two cert files:
 * `certdata.txt`
 * `ca-bundle.crt`

Next, I snagged `keyutil`.

Last, I attempted to create my own `cacert` bundle, but the file it's creating is empty.

Here's my commands:

```bash
# install curl on Ubuntu
sudo apt install curl
# get mk-ca-bundle.pl
wget https://raw.githubusercontent.com/curl/curl/master/lib/mk-ca-bundle.pl
chmod +x mk-ca-bundle.pl
./mk-ca-bundle.pl

# get keyutil
wget https://github.com/use-sparingly/keyutil/releases/download/0.4.0/keyutil-0.4.0.jar
# create a new cacert file
java -jar keyutil-0.4.0.jar --new-keystore cacert --password changeit --import-pem-file ca-bundle.crt
# validate file before deploying
echo "changeit" |keytool -list -keystore cacert
```

Unfortunately, I get the following:
```diff
  Enter keystore password:  Keystore type: JKS
  Keystore provider: SUN

- Your keystore contains 0 entries
```

I assume I'm doing something very basically wrong.

P.S. I can confirm that on Ubuntu 20.04:
* ✅ Firefox loads https://api.insee.fr/catalogue/ just fine
* ✅ openjdk11 (from Canonical repo) loads https://api.insee.fr/catalogue/ just fine.
* 🚫 AdoptOpenJDK11 fails to load https://api.insee.fr/catalogue/

I'll be able to confirm whether or not `cacert` override works once I can get the above working.

**Comment by aahlenst** [comment]
tl;dr: As suspected, the problem is the cacerts shipped with OpenJDK and also AdoptOpenJDK. There are significant less CA certificates in both cacerts keystores. Replacing the bundled cacerts with one generated from Mozilla's list of accepted CAs makes the problem go away.

And now, the details…

OpenJDK version:

```
java -version
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment (build 14.0.1+7)
OpenJDK 64-Bit Server VM (build 14.0.1+7, mixed mode, sharing)
```

AdoptOpenJDK version:

```
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)
```

Sample project to reproduce: https://github.com/aahlenst/cacert-test (JDK 11 and higher). Run `./gradlew check`. Two tests are in it, both should pass.

Number of bundled CA certificates with (Adopt)OpenJDK:

```
keytool -list -keystore jdk-14.0.1/lib/security/cacerts
Warning: use -cacerts option to access cacerts keystore
Enter keystore password:  
Keystore type: JKS
Keystore provider: SUN

Your keystore contains 93 entries
```

Number of CA certificates in Mozilla's list:

```
keytool -list -keystore cacerts-mozilla 
Enter keystore password:  
Keystore type: JKS
Keystore provider: SUN

Your keystore contains 138 entries
```

How I generated cacerts-mozilla:

```bash
$ wget https://hg.mozilla.org/mozilla-central/raw-file/tip/security/nss/lib/ckfw/builtins/certdata.txt .
$ perl mk-ca-bundle.pl -n > ca-bundle.crt
$ java -jar keyutil-0.4.0.jar --import --new-keystore cacerts-mozilla --password changeit --force-new-overwrite --import-pem-file ca-bundle.crt
```

With the bundled cacerts:

```
./gradlew clean check

> Task :test FAILED

SampleClientTest > callSiteWithKnownCa() PASSED

SampleClientTest > callSiteWithUnknownCa() FAILED
    javax.net.ssl.SSLHandshakeException at SampleClientTest.java:16
        Caused by: javax.net.ssl.SSLHandshakeException at Alert.java:131
            Caused by: sun.security.validator.ValidatorException at PKIXValidator.java:439
                Caused by: sun.security.provider.certpath.SunCertPathBuilderException at SunCertPathBuilder.java:141

2 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///root/cacert-test/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
4 actionable tasks: 4 executed
```

After replacing the bundled cacerts with the newly generated one (`cp cacerts-mozilla jdk-14.0.1/lib/security/cacerts`):

```
./gradlew clean check

> Task :test

SampleClientTest > callSiteWithKnownCa() PASSED

SampleClientTest > callSiteWithUnknownCa() PASSED

BUILD SUCCESSFUL in 2s
4 actionable tasks: 4 executed
```

The sites I used to test: https://google.com/ and https://api.insee.fr/catalogue/.

It's expected that JDK's shipped with Ubuntu/Debian work because they use a different cacerts build from Mozilla's list.

Side note: I tested with BellSoft Liberica 13, too. Same problem. 

Personal commentary: I can see how one can argue that it makes sense to ship the same cacerts file as upstream because one of AdoptOpenJDKs objectives is to match OpenJDK as closely as possible. In this case, I think we're doing our users a disservice. We should either delegate what CA certificates are accepted to the operating system or, if this isn't possible, at least ship the best curated list of trusted CA certificates and that's Mozilla's.

**Comment by aahlenst** [comment]
@karianna Please advise how to proceed here. I see three options:

1. Try to persuade OpenJDK to follow Mozilla's lead.
2. Regenerate cacerts at AdoptOpenJDK from Mozilla's list, either on each build or before each release. Might need TSC discussion.
3. Do nothing.

Necessary information to reproduce should be present.

**Comment by karianna** [comment]
Added to TSC-Agenda

**Comment by karianna** [comment]
Discussed at the TSC meeting.  An informal vote was to ask OpenJDK to include these certs for all downstream repos.  @aahlenst can you work with Severin to raise this?

**Comment by jerboaa** [comment]
@aahlenst I'd suggest to raise a discussion on [jdk-dev](https://mail.openjdk.java.net/mailman/listinfo/jdk-dev) first. Gather some input on the proposal.

**Comment by aahlenst** [comment]
Thread: https://mail.openjdk.java.net/pipermail/jdk-dev/2020-May/004305.html

**Comment by jwesselink** [comment]
The certificate that I motioned in #1743 and the "Certigna" certificate motioned in this issue are both present in the OpenJDK docker image (openjdk:11-jre-slim)

@aahlenst Quote from the mailthread: "We ship the same cacerts file as OpenJDK" 
I'm not sure if that true or am I missing something?

**Comment by aahlenst** [comment]
@jwesselink https://github.com/AdoptOpenJDK/openjdk-support/issues/13#issuecomment-626147267 shows what I did based on OpenJDK 14.0.1 and AdoptOpenJDK 14.0.1. It is very well possible that I made a mistake. So it would be great if you could verify it. But please do not bring anything else into the mix like containers. Only compare binary builds without OS integration.

**Comment by karianna** [comment]
The discussion at OpenJDK has been interesting, but I'm not getting a sense that there's broad consensus to change. So I think this goes back to the TSC to see if we add them regardless.

**Comment by karianna** [comment]
TSC has agreed in principle to add the Mozilla certs - @gdams will take a first look at it.

**Comment by breun** [comment]
I ran into this issue today as well. The Dutch government runs its own root CA, but this is not included in AdoptOpenJDK's `cacerts`, while it is included in the `cacerts` file from the `ca-certificates` package on CentOS 8:

```
$ echo changeit | keytool -list -v -keystore $(find $JAVA_HOME -name cacerts) | grep ^Owner | grep Staat
$ echo changeit | keytool -list -v -keystore /etc/pki/ca-trust/extracted/java/cacerts | grep ^Owner | grep Staat
Owner: CN=Staat der Nederlanden EV Root CA, O=Staat der Nederlanden, C=NL
Owner: CN=Staat der Nederlanden Root CA - G2, O=Staat der Nederlanden, C=NL
Owner: CN=Staat der Nederlanden Root CA - G3, O=Staat der Nederlanden, C=NL
```

I'm tempted to just symlink AdoptOpenJDK's `cacerts` to `/etc/pki/ca-trust/extracted/java/cacerts`.

**Comment by breun** [comment]
I also tried Azul Zulu Community, and that OpenJDK distribution does contain this root CA:

```
$ echo changeit | keytool -list -v -keystore $(find $JAVA_HOME -name cacerts) 2> /dev/null | grep ^Owner | grep Staat
Owner: CN=Staat der Nederlanden Root CA - G3, O=Staat der Nederlanden, C=NL
Owner: CN=Staat der Nederlanden EV Root CA, O=Staat der Nederlanden, C=NL
```

Maybe they also use Mozilla's list?

**Comment by aahlenst** [comment]
Sorry it took us so long. 15 should be the first one with a fresh cacerts file. 8 and 11 should get it in October. The nightly builds that appear in the next hours on the website should already have it. Please give them a whirl.
